services/src/util/server.rs

```diff
@@ -294,3 +300,47 @@ pub(crate) fn log_server_info() -> Result<()> {
 pub async fn not_implemented_handler() -> HttpResponse {
     HttpResponse::NotImplemented().finish()
 }
+
+pub struct SocketFd(pub RawFd);
+
+// attach the connections socket file descriptor to the connection data
+// TODO: the socket file descriptor might be re-used by a subsequent connection before we notice that
+//       the connections was closed. We maybe need a global list of all open connections and their socket file descriptors
+//       or some other way to invalidate the socket file descriptor in the old connection when a new one arrives.
+pub fn connection_init(connection: &dyn Any, data: &mut Extensions) {
+    if let Some(sock) = connection.downcast_ref::<TcpStream>() {
+        let fd = sock.as_raw_fd();
+        if let Ok(fd) = NonZeroI32::try_from(fd) {
+            let fd = RawFd::from(fd);
+            data.insert(SocketFd(fd));
+        }
+    }
+}
+
+// start a new task that monitors the request's socket file descriptor and tries to detect when the connection is closed.
+// The returned join handle can be awaited to get notified when the connection is closed.
+// TODO: further verify the heuristic used here to determine the socket status from the `recv` return value
+pub fn connection_closed(req: &HttpRequest) -> Option<JoinHandle<()>> {
+    info!("<< Monitoring req conn >>"); // TODO: remove
+    req.conn_data::<SocketFd>().map(|fd| {
+        let fd = fd.0;
+        crate::util::spawn(async move {
+            let mut data = vec![];
+            // TODO: add a timeout to avoid infinite looping
```

Comment:
can't you tie this to the actual future?

There is a join variant for futures that says: emit the first one that returns… maybe something like this.

Reply:
👍